docs/source/virtual_ecosystem/implementation/hydrology_implementation.md

```diff
@@ -1,29 +1,443 @@
-based on a digital elevation model.
+discharge rate.
+
+The flow direction of water above and below ground is based on a digital elevation model
+which needs to be provided as a NetCDF file at the start of the simulation.
+Here an description of the steps that happen during the hydrology model
+initialisation (plotting only for illustration):
+
+```{code-cell} ipython3
+# # Read elevation data from NetCDF PROBLEM!
+# input_file = "../../../../virtual_ecosystem/example_data/data/example_elevation_data.nc"
+# digital_elevation_model = xr.open_dataset(input_file)
+# elevation = digital_elevation_model['elevation']
+
+# Generate hypothetical digital elevation model data and write in DataArray
+import numpy as np
+from xarray import DataArray
+
+dem_data = np.array(
+  [
+    [1353.0, 583.0, 248.333, 118.0, 24.0, 35.0, 11.0, 46.333, 0.0],
+    [1122.667, 446.111, 404.0, 462.667, 65.444, 52.667, 40.667, 0.0, 11.222],
+    [928.667, 284.778, 277.222, 552.667, 655.111, 671.667, 54.667, 42.222, 831.778],
+    [1008.0, 992.333, 440.0, 582.0, 523.0, 338.333, 596.0, 548.0, 314.0],
+    [619.0, 580.778, 471.222, 271.333, 293.667, 169.0, 609.333, 301.444, 175.667],
+    [374.0, 415.111, 500.111, 318.667, 138.556, 91.444, 88.0, 81.0, 152.778],
+    [1262.0, 316.667, 606.333, 401.0, 116.0, 110.667, 107.0, 16.0, 11.667],
+    [159.333, 1121.778, 1207.222, 524.333, 253.889, 77.444, 76.667, 34.333, 9.889],
+    [0.0, 820.222, 1154.889, 850.333, 299.222, 183.556, 7.333, 8.111, 17.889],
+  ]
+)
+elevation = DataArray(name="elevation", data=dem_data, dims=("x", "y"))
+```
+
+```{code-cell} ipython3
+# Plot the elevation data
+import matplotlib.pyplot as plt
+from matplotlib import colors
+
+plt.figure(figsize=(10, 6))
+elevation.plot(cmap='terrain')
+plt.title('Elevation, m')
+plt.xlabel('x')
+plt.ylabel('y')
+plt.show()
+```
+
+```{code-cell} ipython3
+# Create Grid and Data objects and add elevation data (this happens automatically)
+from virtual_ecosystem.core.grid import Grid
+from virtual_ecosystem.core.data import Data
+
+grid = Grid(grid_type="square", cell_area=8100, cell_nx=9, cell_ny=9)
+data = Data(grid=grid)
+data['elevation'] = elevation
+```
+
+The initialisation step of the hydrology model finds all the neighbours for each grid
+cell and determine which neigbour has the lowest elevation. The code below returns the
+neighbours of the grid cell with `cell_id = 45` as an example.
+
+```{code-cell} ipython3
+grid.set_neighbours(distance=100)
+grid.neighbours[45]
+```
+
+Based on that relationship, the model determines all upstream neighbours
+for each grid cell and creates a drainage map, i.e. a dictionary that contains for each
+grid cell all upstream grid cells:
+
+```{code-cell} ipython3
+from virtual_ecosystem.models.hydrology.above_ground import calculate_drainage_map
+
+drainage_map = calculate_drainage_map(
+  grid=grid,
+  elevation=np.array(data["elevation"]),
+)
+drainage_map
+```
```

Comment:
That's quite a big output and quite technical? Maybe describe one element?

Reply:
agree 👍